src/interface/control_textbox.cpp

```diff
@@ -30,160 +28,233 @@
 // The time for the caret to change visibility (when it flashes).
 static const float CONTROL_CURSOR_TIMER = 0.5f;
 
-ControlTextbox::ControlTextbox() :
+const float ControlTextbox::TEXT_X_OFFSET = 5.f;
+const float ControlTextbox::TEXT_Y_OFFSET = 5.f;
+
+static const size_t UNDO_STACK_SIZE = 10;
+
+ControlTextbox::ControlTextbox(bool multiline) :
   m_validate_string(),
-  m_charlist(),
+  m_multiline(multiline),
+  m_charlist({ {} }),
   m_string(nullptr),
   m_internal_string_backup(""),
+  m_previous_caret_backup(),
+  m_undo_stack(),
+  m_redo_stack(),
   m_cursor_timer(CONTROL_CURSOR_TIMER),
-  m_caret_pos(),
-  m_secondary_caret_pos(),
+  m_caret(),
+  m_secondary_caret(),
   m_shift_pressed(false),
   m_ctrl_pressed(false),
   m_mouse_pressed(),
-  m_current_offset(0)
+  m_scrollbar(),
+  m_offset(0.f, 0.f),
+  m_box_offset(0.f, 0.f),
+  m_box_rect(m_rect)
+{
+  m_scrollbar.reset(new ControlScrollbar(1.f, 1.f, m_offset.y, 35.f));
```

Comment:
`std::make_unique`

also unrelated but i think you can just use `operator=` over reset?